### PR TITLE
Add Rust/c++ indexing to mozilla-beta

### PR DESCRIPTION
**Pull request by staktrace**
The patch queue probably looks scary but 90% of it is extracting bits of the massive `mozilla-central/setup` script into smaller helpers. The only other functional changes in the first n-1 patches are doing more things with `parallel` instead of serially, which helps speed things up. The last patch is the one that updates the release repos to use the helper scripts.

### CODE REVIEW

**Comment by staktrace** [comment]
@emilio no rush but wanted to make sure you saw this

File: mozilla-central/setup [hunk 1]
```diff
@@ -109,40 +109,56 @@ if [ -n "$INDEXED_GIT_REV" ]; then
...
+        # file is small so it's cheap to download and spew to stdout as a check that the analysis data for the platform exists.
         ${CURL} https://index.taskcluster.net/v1/task/gecko.v2.$REVISION.firefox.$PLATFORM-searchfox-debug/artifacts/public/build/target.json ||
         (   echo "WARNING: Unable to find analysis for $PLATFORM for hg rev $INDEXED_HG_REV; skipping analysis merge step for this platform." &&
             continue
         )
 
-        echo "Downloading artifacts for $PLATFORM..."
         TC_PREFIX="https://index.taskcluster.net/v1/task/gecko.v2.${REVISION}.firefox.${PLATFORM}-searchfox-debug/artifacts/public/build"
+        # C++ analysis
+        echo "${CURL} ${TC_PREFIX}/target.mozsearch-index.zip > ${PLATFORM}.mozsearch-index.zip" >> downloads.lst
```
Comment (emilio) [inline mozilla-central/setup]: I think you could just put the urls and do xargs CURL, but there's not a huge difference so fine to leave as is.

File: mozilla-central/setup [hunk 2]
```diff
@@ -105,12 +105,12 @@ if [ -n "$INDEXED_GIT_REV" ]; then
...
 
-    # And now process the downloads
-    for PLATFORM in linux64 macosx64 win64 android-armv7; do
-        date
-        echo "Processing platform $PLATFORM"
-        $CONFIG_REPO/shared/process-tc-artifacts.sh $PLATFORM
-    done
+    # And now process the downloads for each platform in parallel
+    (
+        for PLATFORM in linux64 macosx64 win64 android-armv7; do
```
Comment (emilio) [inline mozilla-central/setup]: Just `echo "linux64\nmacosx64\nwin64\nandroid-armv7"`?

Or `echo "linux64 macosx64 win64 android-armv7" | tr " " "\n"`?

File: mozilla-central/setup [hunk 3]
```diff
@@ -108,17 +108,8 @@ if [ -n "$INDEXED_GIT_REV" ]; then
...
-    while read ANALYSIS_DIR; do
-        mkdir -p "analysis/$ANALYSIS_DIR"
-    done
-    sort analysis-files.list | uniq |
-    while read ANALYSIS_FILE; do
-        echo "Merging analyses for $ANALYSIS_FILE"
-        RUST_LOG=info $MOZSEARCH_PATH/tools/target/release/merge-analyses analysis-*/$ANALYSIS_FILE > analysis/$ANALYSIS_FILE
-    done
-    set -x
+    sort analysis-dirs.list | uniq | parallel --halt now,fail=1 'mkdir -p analysis/{}'
```
Comment (emilio) [inline mozilla-central/setup]: (Not sure how parallel mkdir is possibly helpful, mkdir is generally pretty fast, but sure)

Comment (staktrace) [reply mozilla-central/setup]: Yeah I wasn't sure either, but I figured it couldn't hurt :)

File: mozilla-central/setup (hunk 2 above)
Comment (staktrace) [reply mozilla-central/setup]: Yeah that latter one seems cleaner, will update

**Pull request merged**